scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -920,6 +922,68 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     _.pApply(Sample.fixedSizeGlobally(sampleSize)).map(_.asScala)
   }
 
+  final private class WeightedHeapAggregator(totalWeight: Long, cost: T => Long)(implicit
+    ordering: Ordering[(Long, T)]
+  ) extends Aggregator[T, mutable.PriorityQueue[(Long, T)], Iterable[T]] {
+    @transient private lazy val random = new Random()
+
+    override def prepare(value: T): mutable.PriorityQueue[(Long, T)] = {
+      val queue = mutable.PriorityQueue[(Long, T)]()
+      if (cost(value) <= totalWeight) {
+        queue.enqueue((random.nextLong(), value))
+      }
+      queue
+    }
+
+    override def semigroup: Semigroup[mutable.PriorityQueue[(Long, T)]] = Semigroup.from {
+      (left, right) =>
+        var weight = 0L
+        val queue = mutable.PriorityQueue[(Long, T)]()
+        while (weight < totalWeight && (right.nonEmpty || left.nonEmpty)) {
+          val (w, v) = (left.headOption, right.headOption) match {
+            case (Some((lid, lvalue)), Some((rid, _))) if lid >= rid =>
+              (cost(lvalue), left.dequeue())
+            case (Some(_), Some((_, rvalue))) =>
+              (cost(rvalue), right.dequeue())
+            case (Some((_, lvalue)), _) =>
+              (cost(lvalue), left.dequeue())
+            case (_, Some((_, rvalue))) =>
+              (cost(rvalue), right.dequeue())
+          }
+          weight += w
+          if (weight <= totalWeight) {
+            queue.enqueue(v)
+          }
```

Comment:
In `batchByteSized`, we exceed the size and send the batch. Here we do not add the last element. WDYT is best? We should align the 2 behaviors.

Reply:
IMO it is good enough to see "adding next item would exceed batch size" and sending the batch then. I guess the only edge case is when there is a single element larger than the batch size in which case :shrug: drop it? send it alone?

Looking for "best fit" seems like an overoptimization for the cases this api would be useful

Reply:
agreed. In batch all elements needs to be sent, even when a single element is bigger than the batch size.

In the sample, reduce drops elements until we're bellow the size, so it feels a consistent behavior to not even add it if it exceeds the size.